rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -238,29 +243,145 @@ object Reduce {
+        Task raiseError new Error(
+          "Error: parallel or non expression found where expression expected.")
+      else
+        p.exprs match {
+          case Expr(EVarBody(EVar(v))) +: Nil =>
+            for {
+              p       <- eval(v.get)
+              exprVal <- evalSingleExpr(p)
+            } yield exprVal
+          case (e: Expr) +: Nil => evalExpr(e)
+          case _ =>
+            Task raiseError new Error("Error: Multiple expressions given.")
+        }
+
+    def evalExpr(expr: Expr)(implicit env: Env[Par]): Task[Expr] = {
+      def relop(p1: Par,
+                p2: Par,
+                relopb: (Boolean, Boolean) => Boolean,
+                relopi: (Int, Int) => Boolean,
+                relops: (String, String) => Boolean): Task[Expr] =
+        for {
+          v1 <- evalSingleExpr(p1)
+          v2 <- evalSingleExpr(p2)
+          result <- (v1.exprInstance, v2.exprInstance) match {
+                     case (GBool(b1), GBool(b2))     => Task.pure(GBool(relopb(b1, b2)))
+                     case (GInt(i1), GInt(i2))       => Task.pure(GBool(relopi(i1, i2)))
+                     case (GString(s1), GString(s2)) => Task.pure(GBool(relops(s1, s2)))
+                     case _                          => Task raiseError new Error("Unexpected compare: " + v1 + " vs. " + v2)
+                   }
+        } yield result
+
       expr.exprInstance match {
-        case x: GBool     => Task.pure[Expr](x)
-        case x: GInt      => Task.pure[Expr](x)
-        case x: GString   => Task.pure[Expr](x)
-        case x: GUri      => Task.pure[Expr](x)
-        case EVarBody(ev) => Task.pure[Expr](ev)
+        case x: GBool   => Task.pure[Expr](x)
+        case x: GInt    => Task.pure[Expr](x)
+        case x: GString => Task.pure[Expr](x)
+        case x: GUri    => Task.pure[Expr](x)
+        case ENotBody(ENot(p)) =>
+          for {
+            b <- evalToBool(p.get)
+          } yield GBool(!b)
+        case ENegBody(ENeg(p)) =>
+          for {
+            v <- evalToInt(p.get)
+          } yield GInt(-v)
+        case EMultBody(EMult(p1, p2)) =>
+          for {
+            v1 <- evalToInt(p1.get)
+            v2 <- evalToInt(p2.get)
+          } yield GInt(v1 * v2)
+        case EDivBody(EDiv(p1, p2)) =>
+          for {
+            v1 <- evalToInt(p1.get)
+            v2 <- evalToInt(p2.get)
+          } yield GInt(v1 / v2)
         case EPlusBody(EPlus(p1, p2)) =>
           for {
             v1 <- evalToInt(p1.get)
             v2 <- evalToInt(p2.get)
           } yield GInt(v1 + v2)
-        case _ => Task raiseError new Error("Unimplemented expression.")
+        case EMinusBody(EMinus(p1, p2)) =>
+          for {
+            v1 <- evalToInt(p1.get)
+            v2 <- evalToInt(p2.get)
+          } yield GInt(v1 - v2)
+        case ELtBody(ELt(p1, p2))   => relop(p1.get, p2.get, (_ < _), (_ < _), (_ < _))
+        case ELteBody(ELte(p1, p2)) => relop(p1.get, p2.get, (_ <= _), (_ <= _), (_ < _))
```

Comment:
`relop(p1.get, p2.get, (_ <= _), (_ <= _), (_ <= _))`

Reply:
Done.

Reply:
Scala won't let you express the type that I'd really like to write:
(forall a. (Ord a) => a -> a -> Bool)
Works in haskell with Rank 2 types.